internal/app/machined/pkg/controllers/block/volume_config.go

```diff
@@ -0,0 +1,287 @@
+		default:
+			return fmt.Errorf("unsupported encryption key type: slot %d", key.Slot())
+		}
+	}
+
+	return nil
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *VolumeConfigController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		select {
+		case <-r.EventCh():
+		case <-ctx.Done():
+			return nil
+		}
+
+		// load config if present
+		cfg, err := safe.ReaderGetByID[*config.MachineConfig](ctx, r, config.V1Alpha1ID)
+		if err != nil && !state.IsNotFoundError(err) {
+			return fmt.Errorf("error fetching machine configuration")
+		}
+
+		// load STATE encryption meta key
+		encryptionMeta, err := safe.ReaderGetByID[*runtime.MetaKey](ctx, r, runtime.MetaKeyTagToID(meta.StateEncryptionConfig))
+		if err != nil && !state.IsNotFoundError(err) {
+			return fmt.Errorf("error fetching state encryption meta key")
+		}
+
+		r.StartTrackingOutputs()
+
+		// META volume discovery, always created unconditionally
+		if err = safe.WriterModify(ctx, r,
+			block.NewVolumeConfig(block.NamespaceName, constants.MetaPartitionLabel),
+			func(vc *block.VolumeConfig) error {
+				vc.TypedSpec().Type = block.VolumeTypePartition
+				vc.TypedSpec().Locator = block.LocatorSpec{
+					Match: metaMatch(),
+				}
+
+				return nil
+			},
+		); err != nil {
+			return fmt.Errorf("error creating meta volume configuration: %w", err)
+		}
+
+		// if config is present (v1apha1 part of now)
+		// [TODO]: support custom configuration later
+		configurationPresent := cfg != nil && cfg.Config().Machine() != nil
+
+		// STATE configuration should be always created, but it depends on the configuration presence
+		if err = safe.WriterModify(ctx, r,
+			block.NewVolumeConfig(block.NamespaceName, constants.StatePartitionLabel),
+			func(vc *block.VolumeConfig) error {
+				vc.TypedSpec().Type = block.VolumeTypePartition
+				vc.TypedSpec().Mount = block.MountSpec{
+					TargetPath: constants.StateMountPoint,
+				}
+
+				if configurationPresent {
+					vc.TypedSpec().Provisioning = block.ProvisioningSpec{
+						Wave: block.WaveSystemDisk,
+						DiskSelector: block.DiskSelector{
+							Match: systemDiskMatch(),
+						},
+						PartitionSpec: block.PartitionSpec{
+							MinSize:  partition.StateSize,
+							MaxSize:  partition.StateSize,
+							Label:    constants.StatePartitionLabel,
+							TypeUUID: partition.LinuxFilesystemData,
+						},
+						FilesystemSpec: block.FilesystemSpec{
+							Type:  block.FilesystemTypeXFS,
+							Label: constants.StatePartitionLabel,
+						},
+					}
+
+					vc.TypedSpec().Locator = block.LocatorSpec{
+						Match: labelVolumeMatch(constants.StatePartitionLabel),
+					}
+
+					if err = convertEncryption(
+						cfg.Config().Machine().SystemDiskEncryption().Get(constants.StatePartitionLabel),
+						vc.TypedSpec(),
+					); err != nil {
+						return fmt.Errorf("error converting encryption for %s: %w", constants.StatePartitionLabel, err)
+					}
+				} else {
```

Comment:
I'd suggest to `return nil` from this `if` to have less nesting here:

```suggestion
				if configurationPresent {
					vc.TypedSpec().Provisioning = block.ProvisioningSpec{
						Wave: block.WaveSystemDisk,
						DiskSelector: block.DiskSelector{
							Match: systemDiskMatch(),
						},
						PartitionSpec: block.PartitionSpec{
							MinSize:  partition.StateSize,
							MaxSize:  partition.StateSize,
							Label:    constants.StatePartitionLabel,
							TypeUUID: partition.LinuxFilesystemData,
						},
						FilesystemSpec: block.FilesystemSpec{
							Type:  block.FilesystemTypeXFS,
							Label: constants.StatePartitionLabel,
						},
					}

					vc.TypedSpec().Locator = block.LocatorSpec{
						Match: labelVolumeMatch(constants.StatePartitionLabel),
					}

					if err = convertEncryption(
						cfg.Config().Machine().SystemDiskEncryption().Get(constants.StatePartitionLabel),
						vc.TypedSpec(),
					); err != nil {
						return fmt.Errorf("error converting encryption for %s: %w", constants.StatePartitionLabel, err)
					}
					
					return nil
				}
```

Reply:
I don't think I can `return nil` there, but I see your point that this method is too long, I will see into breaking it apart a bit

Reply:
I refactored into smaller functions